cmd/nodecmd/join.go

```diff
@@ -0,0 +1,478 @@
+	if err != nil {
+		ux.Logger.PrintToUser("failed to check if node is a validator on Primary Network: %s", err)
+	} else if isValidator {
+		return fmt.Errorf("node %s is already a validator on Primary Network", nodeID)
+	}
+	return nil
+}
+
+func addNodeAsPrimaryNetworkValidator(nodeID ids.NodeID, network models.Network) (bool, error) {
+	if err := checkNodeIsPrimaryNetworkValidator(nodeID, network); err == nil {
+		err = validatePrimaryNetwork(nodeID, network)
+		if err != nil {
+			return false, err
+		}
+		ux.Logger.PrintToUser("Node successfully added as Primary Network validator!")
+		return true, nil
+	}
+	return false, nil
+}
+
+func waitForNodeToBePrimaryNetworkValidator(nodeID ids.NodeID) {
+	ux.Logger.PrintToUser("Waiting for the node to start as a Primary Network Validator...")
+	// wait for 20 seconds because we set the start time to be in 20 seconds
+	time.Sleep(20 * time.Second)
+	// long polling: try up to 5 times
+	for i := 0; i < 5; i++ {
+		// checkNodeIsPrimaryNetworkValidator only returns err if node is already a Primary Network validator
+		if err := checkNodeIsPrimaryNetworkValidator(nodeID, models.Fuji); err != nil {
+			break
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func joinSubnet(_ *cobra.Command, args []string) error {
+	clusterName := args[0]
+	err := setupAnsible()
+	if err != nil {
+		return err
+	}
+	isBootstrapped, err := checkNodeIsBootstrapped(clusterName)
+	if err != nil {
+		return err
+	}
+	if !isBootstrapped {
+		return errors.New("node is not bootstrapped yet, please try again later")
+	}
+	nodeIDStr, err := getClusterNodeID(clusterName)
+	if err != nil {
+		return err
+	}
+	nodeID, err := ids.NodeIDFromString(nodeIDStr)
+	if err != nil {
+		return err
+	}
+	if subnetName == "" {
+		// if no subnet is given in the flag, node will only be added as Primary Network Validator
+		_, err = addNodeAsPrimaryNetworkValidator(nodeID, models.Fuji)
+		return err
+	}
+	if _, err = subnetcmd.ValidateSubnetNameAndGetChains([]string{subnetName}); err != nil {
+		return err
+	}
+	sc, err := app.LoadSidecar(subnetName)
+	if err != nil {
+		return err
+	}
+	blockchainID := sc.Networks[models.Fuji.String()].BlockchainID
+	if blockchainID == ids.Empty {
+		return ErrNoBlockchainID
+	}
+	// we have to check if node is synced to subnet before adding the node as a validator
+	isSubnetSynced, err := getClusterSubnetSyncStatus(blockchainID.String(), clusterName)
+	if err != nil {
+		return err
+	}
+	if !isSubnetSynced {
+		return errors.New("node is not synced to subnet yet, please try again later")
+	}
+	addedNodeAsPrimaryNetworkValidator, err := addNodeAsPrimaryNetworkValidator(nodeID, models.Fuji)
+	if err != nil {
+		return err
+	}
+	if addedNodeAsPrimaryNetworkValidator {
+		waitForNodeToBePrimaryNetworkValidator(nodeID)
+	}
+	return addNodeAsSubnetValidator(nodeIDStr, models.Fuji)
+}
+
+// convertToAVAXStr converts nanoAVAX to AVAX
+func convertToAVAXStr(weight uint64) string {
```

Comment:
You should rename this to `convertNanoAvaxToAvaxString` or something like that. Name isn't clear.